File: src/logic/pieces/queen.py
from logic.move import Move
from .piece import Piece

class Queen(Piece):
    def legal_moves(self, board: "Board", / , looking_for_check = False) -> list[Move]:
        ret = []

        # looking north east
        ret.extend(self._look_direction(board, 1, 1))

        # looking south east
        ret.extend(self._look_direction(board, 1, -1))

        # looking south west
        ret.extend(self._look_direction(board, -1, -1))

        # looking north west
        ret.extend(self._look_direction(board, -1, 1))

        # looking east
        ret.extend(self._look_direction(board, 1, 0))

        # looking south
        ret.extend(self._look_direction(board, 0, -1))

        # looking west
        ret.extend(self._look_direction(board, -1, 0))

        # looking north
        ret.extend(self._look_direction(board, 0, 1))

        if not looking_for_check:# and board.is_check_for(self.colour):
            return self.keep_only_blocking(ret, board)

        return ret

File: cpp/src/model/pieces/queen.cpp
#include "../board/board.hpp"
#include "../utils/coords.hpp"
#include "../utils/move.hpp"
#include "piece.hpp"

#include <vector>

std::vector<Move> queen_moves(const Board& b, const Coords xy) {
    std::vector<Move> ret;
    auto e = look_direction(b, xy, 1, 0);
    ret.insert(ret.end(), e.begin(), e.end());

    auto s = look_direction(b, xy, 0, -1);
    ret.insert(ret.end(), s.begin(), s.end());

    auto w = look_direction(b, xy, -1, 0);
    ret.insert(ret.end(), w.begin(), w.end());

    auto n = look_direction(b, xy, 0, 1);
    ret.insert(ret.end(), n.begin(), n.end());

    auto ne = look_direction(b, xy, 1, 1);
    ret.insert(ret.end(), ne.begin(), ne.end());

    auto se = look_direction(b, xy, 1, -1);
    ret.insert(ret.end(), se.begin(), se.end());

    auto sw = look_direction(b, xy, -1, -1);
    ret.insert(ret.end(), sw.begin(), sw.end());

    auto nw = look_direction(b, xy, -1, 1);
    ret.insert(ret.end(), nw.begin(), nw.end());

    return ret;
}
